### PR TITLE
PyBind `GoldenTensor` w/ Integral Types (#1294)

### LINKED ISSUE #1294: Support multiple data types for golden comparison
The groundwork and top level API for doing golden comparison in tt-mlir has been laid as part of this PR:
- https://github.com/tenstorrent/tt-mlir/pull/1218

As follow on work we'd like to support golden comparison for any datatype, per comments on the PR:
1. https://github.com/tenstorrent/tt-mlir/pull/1218#discussion_r1844453209
2. https://github.com/tenstorrent/tt-mlir/pull/1218#discussion_r1841489612

Both of the above comments seem like possible approaches to solving the issue.  I think we should aim to try and do option 2, which seems more robust and efficient, copied here:

We could define a pybind [buffer protocol](https://pybind11.readthedocs.io/en/stable/advanced/pycpp/numpy.html)

```c++
py::class_<::tt::target::GoldenTensor>(m, "GoldenTensor", py::buffer_protocol())
   .def_buffer([](GoldenTensor const *t) -> py::buffer_info {
        return py::buffer_info(
            t->data()->data(),                      /* Pointer to buffer */
            sizeof(float),                          /* Size of one scalar */
            py::format_descriptor<float>::format(), /* Python struct-style format descriptor */
            t->shape()->size(),                     /* Number of dimensions */
            t->shape(),                             /* Buffer dimensions */
            t->strides()                            /* Strides (in bytes) for each index */
        );
    });

// get_debug_info_golden returns a ::tt::target::GoldenTensor
```

On the python side we can use [torch.frombuffer(golden_tensor)](https://pytorch.org/docs/stable/generated/torch.frombuffer.html).

### PR DESCRIPTION
**Pull request by ctodTT**
This change introduces a PyBound class `GoldenTensor` that reflects the same type from TTRT. This binding supports the following element types for a `GoldenTensor`:
  - uint8
  - uint16
  - uint32
  - float32

Closes #1294

### CODE REVIEW

File: runtime/tools/python/ttrt/common/golden.py [hunk 1]
```diff
@@ -127,27 +128,40 @@ def golden(binary, programContext, opContext):
...
+            print("Golden tensor is None - skipping golden comparison")
             return
 
         if len(op_output_tensor) == 0:
             print("Output tensor is empty - skipping golden comparison")
             return
 
-        if len(op_golden_tensor) != len(op_output_tensor):
+        # Determine the element type of the golden tensor
+        match op_golden_tensor.dtype:
```
Comment (nsmithtt) [inline runtime/tools/python/ttrt/common/golden.py:140]: So technically on macOS we only support python 3.9, but then again this code will only run on runtime enabled builds i.e. linux which has python 3.10.  We'll see what CI says :)

File: runtime/tools/python/ttrt/common/run.py [hunk 2]
```diff
@@ -407,10 +409,26 @@ def _execute(binaries):
...
+                                        case DataType.UInt32:
+                                            dtype = torch.uint32
+                                        case DataType.UInt16:
+                                            dtype = torch.uint16
+                                        case DataType.UInt8:
+                                            dtype = torch.uint8
+                                        case _:
+                                            raise ValueError(
+                                                "Only F32 and unsigned integers are supported for `GoldenTensor`s"
+                                            )
```
Comment (nsmithtt) [inline runtime/tools/python/ttrt/common/run.py:426]: Maybe we should have a common utility function that can do this conversion?

**Review by nsmithtt — APPROVED** [review]
This looks great!

File: runtime/tools/python/ttrt/common/golden.py (hunk 1 above)
Comment (ctodTT) [reply runtime/tools/python/ttrt/common/golden.py]: CI seems fine with it, if we run into an issue we can switch this to a map or something.

File: runtime/tools/python/ttrt/common/run.py (hunk 2 above)
Comment (ctodTT) [reply runtime/tools/python/ttrt/common/run.py]: ✅

File: runtime/tools/python/ttrt/runtime/module.cpp [hunk 3]
```diff
@@ -136,4 +136,16 @@ PYBIND11_MODULE(_C, m) {
...
+   * Cleanup code to force a well ordered destruction w.r.t. the GIL
+   *
+   * TODO: `clangd` reports that `capsule` is deprecated, figure out how to do
+   * this canonically
+   */
+  static int unused; // the capsule needs something to reference
+  py::capsule cleanup(&unused, [](PyObject *) {
+    ::tt::runtime::debug::Hooks::get().unregisterHooks();
+  });
+  m.add_object("_cleanup", cleanup);
```
Comment (nsmithtt) [inline runtime/tools/python/ttrt/runtime/module.cpp:150]: Re deprecated, it looks like only this variation of the capsule constructor is deprecated.  It looks like we can avoid the deprecated constructor and more succinctly written ([from their official docs](https://pybind11.readthedocs.io/en/stable/advanced/misc.html#module-destructors)):
```c++
auto cleanup_callback = []() {
    ::tt::runtime::debug::Hooks::get().unregisterHooks();
};
m.add_object("_cleanup", py::capsule(cleanup_callback));
```

**Review by nsmithtt — APPROVED** [review]
Minor comment about deprecated constructor, other than that let's land it :)

File: runtime/include/tt/runtime/detail/debug.h [hunk 4]
```diff
@@ -63,14 +63,17 @@ struct Hooks {
 #endif
   }
 
+  void unregisterHooks() const { operatorCallback = std::nullopt; }
```
Comment (nsmithtt) [inline runtime/include/tt/runtime/detail/debug.h:66]: Oh there's also a build failure:
```
/__w/tt-mlir/tt-mlir/runtime/include/tt/runtime/detail/debug.h:66:34: error: use of undeclared identifier 'operatorCallback'
   66 |   void unregisterHooks() const { operatorCallback = std::nullopt; }
      | 
```

`operatorCallback` is only defined for debug builds.  Probably the right fix is:
```c++
void unregisterHooks() const {
#if defined(TT_RUNTIME_DEBUG) && TT_RUNTIME_DEBUG == 1
  operatorCallback = std::nullopt;
#endif
}
```

**Review by tapspatel — APPROVED** [review]

**Review by kmabeeTT — APPROVED** [review]
Thanks, Collin!

**Comment by tapspatel** [comment]
Context for revert: @ctodTT and I noticed that using the python file line and column as the location results in many of the decomposed ops from TTIR to have that loc appended to it, while use a global id does not (we don't know why yet). This results in ops that shouldn't be doing golden eval (eg to_device op) doing golden eval on the callback. The commit has been reverted while this issue is investigated.

**Pull request merged**